docs/compatibility.md

```diff
@@ -0,0 +1,183 @@
+# StableHLO Compatibility
+
+## Compatibility Guarantees
+
+(G1) libStablehlo provides 6 months of backward compatibility, i.e. StableHLO programs serialized by an old version of libStablehlo have the same semantics when deserialized by new versions of libStablehlo within the compatibility window.
```

Comment:
Interesting. Let me think more on this. Maintaining all "Too old" errors may require some overhead, we would probably need relics living in the codebase for the purpose of erroring. Maybe that's ok though. 

If we know the bytecode version, we may be able to display a warning up front before attempting to deserialize. "This bytecode file is from an unsupported version. Compatibility is not guaranteed." 

Currently, for renaming changes for example, the error would most likely be "unrecognized op ..." or "missing attribute ...".

This is all similar for the forwards compatibility concern below.

Reply:
I'm unsure if Sean is asking about the error message, or guaranteeing to throw an error. 

I think it's actually critical to properly throw an error instead of having a possibility for a new library to load an old serialized StableHLO and handle it incorrectly & silently. Is this guaranteed? On top of this, a proper error message is nice to have. 

To put it in another way, in-compatibility detection & error handling is as important as compatibility in my opinion. We don't want developers who use a one-year old serialized file to fail silently and wonder why.

Reply:
MLIR bytecode natively has a header to encode the producer version string, so we can support any kind of scheme and diagnostic on load.

Reply:
A serialized file can deterministically compute valid consumer versions during deserialization and will error if consumer version is invalid. This applies to files that are "too old" and "too new" for a given consumer.